src/WebJobs.Script/Host/AggregateFunctionMetadataProvider.cs

```diff
@@ -17,44 +17,69 @@
 
 namespace Microsoft.Azure.WebJobs.Script
 {
-    public class WorkerFunctionMetadataProvider : IFunctionMetadataProvider
+    public class AggregateFunctionMetadataProvider : IFunctionMetadataProvider
     {
         private readonly Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
         private readonly ILogger _logger;
         private readonly IFunctionInvocationDispatcher _dispatcher;
         private ImmutableArray<FunctionMetadata> _functions;
+        private IFunctionMetadataProvider _hostFunctionMetadataProvider;
 
-        public WorkerFunctionMetadataProvider(ILogger<WorkerFunctionMetadataProvider> logger, IFunctionInvocationDispatcher invocationDispatcher)
+        public AggregateFunctionMetadataProvider(
+            ILogger<AggregateFunctionMetadataProvider> logger,
+            IFunctionInvocationDispatcher invocationDispatcher,
+            IFunctionMetadataProvider hostFunctionMetadataProvider)
         {
             _logger = logger;
-            _dispatcher = invocationDispatcher ?? throw new ArgumentNullException(nameof(invocationDispatcher));
+            _dispatcher = invocationDispatcher;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
             IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
+
+            bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
+
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
-                if (_dispatcher != null)
+                if (_dispatcher == null && workerIndexing == true)
+                {
+                    throw new ArgumentNullException(nameof(_dispatcher));
```

Comment:
What happens if dispatcher is null and worker indexing is false? Do we also throw? Just wondering if  throwing `ArgumentNullException` is better were it was originally in the constructor, it would also simplify this function?

Reply:
I have cleaned up the code a bit here. The reason I cannot throw ArgumentNullException in constructor is - if workerIndexing = false, dispatcher can be null but if workerIndexing = true then it can't be and both of these scenarios are being handled in this Aggregate class.